### PR TITLE
Flink: Adds support for Flink 1.19 version

### PR DESCRIPTION
**Pull request by rodmeneses**
This is how I prepared this PR:
1. `cd flink`
2. `git mv v.18/ v1.19`
3. `git commit -am"Flink: Move flink/v1.18 to flink/v1.19"`
4. `rm -rf v1.18/`
5. `cp -R v1.19/ v1.18`
6. `git add v1.18/`
7. `git commit -am"Flink: Recover flink/1.18 files from history"`
8. code changes, properties changes so that all unit test pass
9.  `git commit -am"Flink: Refactoring code and properties to make Flink 1.19 to work"`

*a new PR for deleting v1.16 will immediately follow up after this is merged*

### CODE REVIEW

**Comment by rodmeneses** [comment]
cc: @pvary @stevenzwu @mas-chen please take a look at your earliest convenience. Thanks

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java [hunk 1]
```diff
@@ -126,4 +126,18 @@ protected void dropCatalog(String catalogName, boolean ifExists) {
...
+  /**
+   * We can not drop currently used database after FLINK-33226, so we have make sure that we do not
+   * use the current database before dropping it. This method creates a database called 'temp', uses
+   * it and drops the one requested.
+   *
+   * @param database The database to drop
+   * @param ifExists If we should use the 'IF EXISTS' when dropping the database
+   */
+  protected void dropDatabase(String database, boolean ifExists) {
+    sql("CREATE DATABASE IF NOT EXISTS temp");
```
Comment (pvary) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/FlinkTestBase.java:139]: Do we have a default database to use? I find it strange that we are creating a database to drop a database 😄 
What happens in Flink if we do not create/use a database before creating a table?

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ public void before() {
   @Override
   public void clean() {
     sql("DROP TABLE IF EXISTS %s", TABLE_NAME);
-    sql("DROP DATABASE IF EXISTS %s", DATABASE_NAME);
+    dropDatabase(DATABASE_NAME, true);
```
Comment (pvary) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java:101]: Do we really need this? Something behind the scenes issue a `USE DATABASE` command?

If so, this could be a breaking change which we might want to highlight, as creating and dropping a database immediately after that won't work anymore

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/enumerator/TestContinuousSplitPlannerImpl.java [hunk 3]
```diff
@@ -443,9 +446,10 @@ public void testIncrementalFromSnapshotTimestampWithInvalidIds() throws Exceptio
     ContinuousSplitPlannerImpl splitPlanner =
         new ContinuousSplitPlannerImpl(
             tableResource.tableLoader().clone(), scanContextWithInvalidSnapshotId, null);
+
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/enumerator/TestContinuousSplitPlannerImpl.java:449]: seems like unnecessary changes here

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestStreamingMonitorFunction.java [hunk 4]
```diff
@@ -248,6 +248,7 @@ public void testInvalidMaxPlanningSnapshotCount() {
             .monitorInterval(Duration.ofMillis(100))
             .maxPlanningSnapshotCount(0)
             .build();
+
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestStreamingMonitorFunction.java:251]: unnecessary change

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkInputFormatReaderDeletes.java [hunk 5]
```diff
@@ -35,6 +35,7 @@
 import org.apache.iceberg.util.StructLikeSet;
 
 public class TestFlinkInputFormatReaderDeletes extends TestFlinkReaderDeletesBase {
+
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkInputFormatReaderDeletes.java:38]: I think it would be good to remove all of these whitespace changes, which would make it much easier to review the PR

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java [hunk 6]
```diff
@@ -294,39 +308,284 @@ public void testAlterTable() throws TableNotExistException {
...
+            new Schema(
+                    Types.NestedField.optional(1, "id", Types.LongType.get()),
+                    Types.NestedField.optional(2, "dt", Types.StringType.get()),
+                    Types.NestedField.optional(
+                        3, "col1", Types.StringType.get(), "comment for col1"),
+                    Types.NestedField.optional(4, "col2", Types.LongType.get()))
+                .asStruct());
+    // Adding a required field should fail because Iceberg's SchemaUpdate does not allow
+    // incompatible changes.
+    Assertions.assertThatThrownBy(() -> sql("ALTER TABLE tl ADD (pk STRING NOT NULL)"))
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:345]: can be statically imported, similar to `assertThat()`

**Comment by ajantha-bhat** [comment]
@rodmeneses: Just curious to know what command you have used for the step "Flink: Recover flink/1.18 files from history" ? 
It is really nice. 

I thought the only way was to add back the folder (which shows file was added).

**Comment by rodmeneses** [comment]
> @rodmeneses: Just curious to know what command you have used for the step "Flink: Recover flink/1.18 files from history" ? It is really nice.
> 
> I thought the only way was to add back the folder (which shows file was added).

Thanks @ajantha-bhat . I updated the PR description with the steps I took. Please take a look! 😄

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/enumerator/TestContinuousSplitPlannerImpl.java (hunk 3 above)
Comment (rodmeneses) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/enumerator/TestContinuousSplitPlannerImpl.java:449]: HI @nastra and thanks a lot for your reviews.
Some comments:
1. The best way to review this PR is by looking at individual commits. The most important one is the 4th commit called "Flink: Refactoring code and properties to make Flink 1.19 to work". If you review that particular one, you will notice that I'm not introducing the extra line changes.
2. The reason you're seeing those changes is because of the way Github is presenting the differences, which is showing the complete set of differences in the whole PR, and somehow, it's showing diff between the new 1.19 code and the old 1.16 code
3. Now, we do we even have those changes? This is because of a previous back port of changes between versions. There were some inconsistencies when the back port was made and on those old commits the extra lines (or the use of `Assertions.assertThatThrownBy` instead of statically imported) were introduced.
4. So, again, these changes you're referring to are not part of this PR. 
5. I updated the PR description with the exact steps I take to create the PR

Thanks a lot!

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestStreamingMonitorFunction.java (hunk 4 above)
Comment (rodmeneses) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestStreamingMonitorFunction.java:251]: please see my comment above

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkInputFormatReaderDeletes.java (hunk 5 above)
Comment (rodmeneses) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkInputFormatReaderDeletes.java:38]: please see my comment above

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java (hunk 6 above)
Comment (rodmeneses) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java:345]: please see my comment above

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java (hunk 2 above)
Comment (rodmeneses) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java:101]: After a discussion offline with @pvary, we have decided to create a new PR with a better cleaning up logic that will be applied to the unit tests. After that is merged, we can continue with this PR. 
Thanks!

Comment (rodmeneses) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java:101]: I ended up changing this implementation to:
```
  protected void dropDatabase(String database, boolean ifExists) {
    sql("USE CATALOG %s", DEFAULT_CATALOG_NAME);
    sql("USE %s", DEFAULT_DATABASE_NAME);
    sql("DROP DATABASE %s %s", ifExists ? "IF EXISTS" : "", database);
  }
```
which is similar to our current `dropCatalog` implementation.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java [hunk 7]
```diff
@@ -69,7 +69,9 @@ public class FlinkCatalogFactory implements CatalogFactory {
   public static final String HIVE_CONF_DIR = "hive-conf-dir";
   public static final String HADOOP_CONF_DIR = "hadoop-conf-dir";
   public static final String DEFAULT_DATABASE = "default-database";
-  public static final String DEFAULT_DATABASE_NAME = "default";
+  public static final String DEFAULT_DATABASE_NAME = "default_database";
```
Comment (rodmeneses) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:72]: This is needed for 
```
TestIcebergConnector.testCatalogDatabaseConflictWithFlinkDatabase
```
that unit test drop a database, under this condition:
```
if (!isDefaultDatabaseName()) {
        sql("DROP DATABASE `%s`", databaseName());
      }
```
Now, in this unit test context, there only exists one catalog (`default_catalog`) and one database (`default_database`). So, it is impossible to drop that DB because of https://github.com/apache/flink/pull/23501
Also, I noticed that when you switch to `default_catalog` and you try to get the list of databases, the database returned has name `default_database` and not `default`

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestIcebergConnector.java [hunk 8]
```diff
@@ -257,20 +257,12 @@ public void testCreateTableUnderDefaultDatabase() {
...
-              () -> sql("CREATE TABLE `default_catalog`.`%s`.`%s`", databaseName(), TABLE_NAME))
-          .isInstanceOf(org.apache.flink.table.api.TableException.class)
-          .hasMessageStartingWith("Could not execute CreateTable in path");
-    } finally {
-      sql("DROP TABLE IF EXISTS `%s`.`%s`", databaseName(), TABLE_NAME);
-      if (!isDefaultDatabaseName()) {
-        sql("DROP DATABASE `%s`", databaseName());
-      }
-    }
+    testCreateConnectorTable();
```
Comment (rodmeneses) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/TestIcebergConnector.java:260]: Up to this point in the unit test, there's only one catalog `default_catalog` created and only one database `default_database` created so far. So, trying to drop that database will fail, as we are currently using it, and because if `FLINK-33226`, it will be impossible to drop the database.

File: flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkSource.java [hunk 9]
```diff
@@ -60,7 +60,7 @@ protected List<Row> runWithFilter(Expression filter, String sqlFilter, boolean c
   protected List<Row> runWithOptions(Map<String, String> options) throws Exception {
     FlinkSource.Builder builder = FlinkSource.forRowData();
     Optional.ofNullable(options.get("case-sensitive"))
-        .ifPresent(value -> builder.caseSensitive(Boolean.parseBoolean(value)));
+        .ifPresent(value -> builder.caseSensitive(Boolean.getBoolean(value)));
```
Comment (nastra) [inline flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkSource.java:63]: parseBoolean is correct here (this has just been recently changed on main)

**Review by pvary — APPROVED** [review]
@nastra: Any comments? I would like to merge this soon, as any merge to Flink code path will make this PR stale, and @rodmeneses needs to recreate the whole PR.

Thanks,

**Comment by manuzhang** [comment]
Why do we remove Flink 1.1.6 in this PR?

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 10]
```diff
@@ -391,17 +391,16 @@ public void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
               + "an iceberg catalog, Please create table with 'connector'='iceberg' property in a non-iceberg catalog or "
               + "create table without 'connector'='iceberg' related properties in an iceberg table.");
     }
-
-    createIcebergTable(tablePath, table, ignoreIfExists);
+    Preconditions.checkArgument(table instanceof ResolvedCatalogTable, "table should be resolved");
```
Comment (nastra) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:394]: is this new code that's required for Flink 1.19?

Comment (nastra) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:394]: it's not clear to me why there's a diff on `FlinkCatalog`

Comment (nastra) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:394]: I think the problem is that the diff is still from a previous attempt where changes were done based on Flink 1.16. What about removing Flink 1.16 building from the gradle files but doing the actual remove of Flink 1.16 folders in a separate PR? I think that would avoid these diffs here

**Comment by nastra** [comment]
> @nastra: Any comments? I would like to merge this soon, as any merge to Flink code path will make this PR stale, and @rodmeneses needs to recreate the whole PR.
> 
> Thanks,

I think there's still an issue as there are a bunch of files/diffs that are because Flink 1.16 is being removed and git detects it as a move (with some additional changes). This can also be seen when looking at the file path, where a Flink 1.16 file is moved to a Flink 1.19 file, while also adding some diffs where it's not clear why the diff is there in the first place.

My suggestion would be to do the actual removal of the 1.16 directory as a separate PR in an immediate follow-up. This would mean to skip tests 8 + 9 from the PR description, but it's fine to update gradle files to not build 1.16 anymore. Thoughts on the suggestion?

**Comment by rodmeneses** [comment]
> > @nastra: Any comments? I would like to merge this soon, as any merge to Flink code path will make this PR stale, and @rodmeneses needs to recreate the whole PR.
> > Thanks,
> 
> I think there's still an issue as there are a bunch of files/diffs that are because Flink 1.16 is being removed and git detects it as a move (with some additional changes). This can also be seen when looking at the file path, where a Flink 1.16 file is moved to a Flink 1.19 file, while also adding some diffs where it's not clear why the diff is there in the first place.
> 
> My suggestion would be to do the actual removal of the 1.16 directory as a separate PR in an immediate follow-up. This would mean to skip tests 8 + 9 from the PR description, but it's fine to update gradle files to not build 1.16 anymore. Thoughts on the suggestion?

Hi @nastra thanks for your review and comment,
> I think there's still an issue as there are a bunch of files/diffs that are because Flink 1.16 is being removed and git detects it as a move (with some additional changes).
This is because you are looking at the changes in the whole PR. If you se each of the 4 commits individually, you'll find that everything is making sense

I did the approach of not deleting v1.16 and this PR was updated. But, if you see the changes as a whole PR, now it's even worse than before, because we dont see the history properly. 

Given this, I'd suggest to move forward with deleting v1.16 in this same PR. 
Thoughts ? @nastra @pvary

**Review by nastra — APPROVED** [review]
I checked the individual commits and they do what I was expecting, so LGTM. Thanks for updating this. @pvary and I also just talked offline about doing the deletion of Flink 1.16 in a separate PR.

**Comment by pvary** [comment]
Merged to main.
Thanks for the PR @rodmeneses and @nastra for the review!
